module support

wingyplus opened the pull request:
## Tasks

- [x] Every library in the SDK must be netstandard 2.0!
- [x] Generate necessary attributes and interfaces for user code.
- [x] Generate registration method to convert user code into `Dagger.Module` for registering to the Dagger Engine.
    - We need to traverse from root class (`Potato.Potato` for example) and generate child type from their.
- [ ] Generate invoke function.
- [ ] Make `Dagger.SDK` support JSON serialization
    - We need it for returns the value when function get call.
    - For example, `Dagger.Container` must serialize into `Dagger.ContainerID.Value` and deserialize it back. 
    - We can done this in `Dagger.SDK.SourceGenerator` by implementing json converter. 
- [ ] Remove my potato. 🥔 

## Idea on the next PR
- [ ] Verify that namespace and class name must match with dagger module name to make sure we have only single entrypoint.
    - I still doubt on how to do this at the moment. Maybe we can inject some configuration when `Codegen`  get call.
- [ ] ~Remove `Dagger.SDK.Mod`?~
- [ ] Implement project template. Maybe https://learn.microsoft.com/en-us/dotnet/core/tutorials/cli-templates-create-project-template can help us?

Closes #3

wingyplus committed:
feat: module support

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: support generating code

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

pjmagee commented on sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:
```diff
@@ -320,7 +320,7 @@ private static string RenderArgumentValue(InputValue arg, bool addVarSuffix = fa
         }
 
-        throw new Exception($"The type {arg.Type.OfType.Kind} should not be enter here.");
+        throw new Exception($"The argument {arg} should not be enter here.");
```
should not be _entered_

pjmagee commented on main.go:
```diff
@@ -2,15 +2,131 @@
+func (m *DotnetSdk) WithProject(subpath string, modName string) *DotnetSdk {
+	name := strcase.ToCamel(modName)
+	m.Container = m.Container.
+		WithExec([]string{"dotnet", "new", "console", "--framework", "net8.0", "--output", name, "-n", name}).
```
I wonder if there's a way to allow the user, to pick an SDK on their machine. 

dotnet --list-sdks will provide a list of installed SDKS (not all users may be able to use .NET 8, maybe they are only allowed LTS versions of SDK).

pjmagee replied:
I don't think there's an easy way to configure the framework version. How would this look in future, would we bump SDK code here when .NET 9 comes out, we make .NET 9?

pjmagee replied:
I am just thinking here outloud...

Okay so the dagger code will gen .NET 8, imagine I am developer on work machine, and there is a work policy - only use .NET LTS version. If I am to code against APIs, I will modify the generated code to be .NET 6 - maybe I manually downgrade it. In this case, everything should still work because the .NET source Generator is using .NET Standard APIs which is not unique to .NET 8

The only risk we have here is that the OTHER project, may host .NET8 API code we use or language features only available in .NET 8 - in this case, they may need to modify some bits of the code, but in general, the code should be usable across different .NET versions (except .NET FRamework) - So it will likely be compatible with .NET 6,7,8,9 (6 is soon coming to out of support and .NET 8 is LTS)

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core So we are safe here.

wingyplus replied:
I thihnk we should make the `Dagger.SDK` to use `netstandard2` instead of `net8` to make sure the API compatible with the newer version? We already done that in source generator, the SDK should not be hard. And let the user code be .net 8, 9, etc. depends on our dotnet version defined in the docker image.

pjmagee replied:
I think I agree, except for... https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/

I think each year, we can revisit and if Microsoft upgrade Roslyn Source Generators, we can then come back to this and make all `.NET<version>`

ALSO, https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md

I think we need to update Source Generator to use Incremental. I think its minor work, seems to be around performance and not "regenerate" each build time unnecessarily.

wingyplus replied:
Agree 👍

wingyplus replied:
The Dagger.SDK is now netstandard2.0 on main branch. :)

pjmagee commented:
This is looking exciting 👍

wingyplus committed:
feat: implement module runtime

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
demo: add potato example module

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus commented:
UPDATE: I just getting register module and simple invoke works in `potato` example. I also hit a bug that the SDK cannot retrieve array result, cause we cannot inject input arguments into our method correctly.

wingyplus commented on potato/dagger/Potato/Program.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        var potato = JsonSerializer.Deserialize<Potato.Potato>(ParentJson.Value);
+        Debug.Assert(potato != null, nameof(potato) + " != null");
+        return await potato.Echo("C#");
```
I just hit a bug that program got crash when asking input args. So I do temporary hard code argument here to make it works.

wingyplus committed:
feat: register method from reflection

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix merge conflict broken

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix the code to not use id and move module code to new package

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix warnings

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: introduce Dagger.SDK.Mod

Dagger.SDK.Mod hold the logic for construct and invoke function. Current
implmentation using `System.Reflection` to scan class and method.

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: call function work!

And fix bug return list type broken.

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: introduce Dagger.SDK.Mod.SourceGenerator

Introduce a library to uses for generate sources via Source Generator.

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix invalid import

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix incorrect import again

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: update Potato.sln

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: ensure namespace setting is correct

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: filter only class that annotated with Dagger.SDK.Mod.Object

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: try using ForAttributeWithMetadataName

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
I get it!!!!

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: injecting mod attributes and interfaces from source generator

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: convert my potato to use Dagger.SDK.Mod.SourceGenerator

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
chore: add question note

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus commented on sdk/Dagger.SDK.Mod/Dagger.SDK.Mod.csproj:
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
If the source generator plan is going well, we don't need this package anymore. 🎉

wingyplus commented on sdk/Dagger.SDK.Mod.SourceGenerator.Tests/SourceGeneratorTests.cs:
```diff
@@ -0,0 +1,100 @@
+namespace Dagger.SDK.Mod.SourceGenerator.Tests;
+
+[TestClass]
+public class SourceGeneratorTests : VerifyBase
```
I have discussion with Microsoft team in C# discord. They're recommended me to use https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing#supportedframeworks-body-tab rather than `Verify`. I don't see much example from it but it would be nice if we strict with the Microsoft tools.

pjmagee replied:
This is what I thought I had added here: https://github.com/wingyplus/dagger-dotnet-sdk/blob/main/sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator.Tests/Dagger.SDK.SourceGenerator.Tests.csproj#L17

wingyplus commented on sdk/Dagger.SDK.Mod.SourceGenerator/SourceGenerator.cs:
```diff
@@ -0,0 +1,124 @@
+    {
+        context.RegisterPostInitializationOutput(postInitializationContext =>
+        {
+            postInitializationContext.AddSource("Dagger.SDK.Mod_Attributes.g.cs", ModuleAttributesSource());
+            postInitializationContext.AddSource("Dagger.SDK.Mod_Interfaces.g.cs", ModuleInterfacesSource());
```
I'm current stuck with the unit tests because in unit test, it has no reference to the `Dagger.SDK.Mod`, cause it to fail because it cannot find any class related to our SDK. That's why I put it to generator.

pjmagee replied:
I think its difficult to convert to source generator once starting with reflection. It makes you need to change how to design/solution with TDD from the beginning

wingyplus committed:
chore: move generate sources to a new class

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: render register object type method

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix dotnet format and refactor ToObjectTypeDef

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus commented on sdk/Dagger.SDK.Mod.SourceGenerator/SourceGenerator.cs:
```diff
@@ -97,20 +98,12 @@ private static bool IsPartialClass(SyntaxNode node, CancellationToken token)
+            .Functions
+            .Select(static m => m.Name)
+            .Select(static methodName => $"""
+                                          .WithFunction("{methodName}", dag.TypeDef().WithKind(TypeDefKind.STRING)))
```
I committed the syntax bug. 🙈 

will fix it tonight

wingyplus committed:
fix: make type registration works

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: my potato dagger functions back to works

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
feat: move the logic parts in Main program to Entrypoint

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: use Verify.SourceGenerators

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: add more tests

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: reduce size of snapshot files

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: reduce test size

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
test: drop Verify

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
chore: format code

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix test on unix

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: make it compatible with Dagger 0.13+

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
fix: fix cannot list a function

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>

wingyplus committed:
refactor: move fetch function call to Dagger.SDK.Mod project

Signed-off-by: Thanabodee Charoenpiriyakij <wingyminus@gmail.com>
